src/py42/modules/securitydata.py

```diff
@@ -169,6 +178,82 @@ def search_file_events(self, query):
         file_event_client = self._microservices_client_factory.get_file_event_client()
         return file_event_client.search(query)
 
+    def _search_by_hash(self, hash, type):
+        query = FileEventQuery.all(type.eq(hash))
+        response = self.search_file_events(query)
+        return response[u"fileEvents"]
+
+    def _find_file_versions(self, md5_hash, sha256_hash):
+        file_event_client = self._microservices_client_factory.get_file_event_client()
+        pds_client = self._microservices_client_factory.get_preservation_data_service_client()
+        response = file_event_client.get_file_location_detail_by_sha256(sha256_hash)
+
+        if u"locations" not in response and not len(response[u"locations"]):
+            raise Py42ArchiveFileNotFoundError(u"PDS service can't find requested file.")
+
+        for device_id, paths in _parse_file_location_response(response):
+            try:
+                yield pds_client.find_file_versions(md5_hash, sha256_hash, device_id, paths)
+            except Py42HTTPError as err:
+                # API searches multiple paths to find the file to be streamed, as returned by
+                # 'get_file_location_detail_by_sha256', hence we keep looking until we find a stream
+                # to return
+                debug.logger.warning(u"PDS fetch file version API failed. Error: ", err)
+                pass
+
+    def _stream_file(self, file_generator):
+        for response in file_generator:
+            if response.status_code == 204:
+                continue
+            try:
+                storage_node_client = self._microservices_client_factory.create_storage_preservation_client(
+                    response[u"storageNodeURL"]
+                )
+                token = storage_node_client.get_download_token(
+                    response[u"archiveGuid"], response[u"fileId"], response[u"versionTimestamp"],
+                )
+                return storage_node_client.get_file(str(token))
+            except HTTPError as err:
+                # API searches multiple paths to find the file to be streamed, as returned by
+                # 'get_file_location_detail_by_sha256', hence we keep looking until we find a stream
+                # to return
+                debug.logger.warning(u"PDS stream file token API failed, Error:", err)
+                pass
+        raise Py42Error(u"No file available for download.")
```

Comment:
`No file with has <hash> available for download on any storage node`.

Reply:
Done.